### ISSUE TITLE
Feature Request: reduce memory requirements for llama-perplexity on large context sizes (from O(ctx*vocab) to O(vocab))

### ISSUE THREAD

**Issue by andreesteve** [ggml-org/llama.cpp#28733]
### Prerequisites

- [x] I am running the latest code. Mention the version if possible as well.
- [x] I carefully followed the [README.md](https://github.com/ggml-org/llama.cpp/blob/master/README.md).
- [x] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [x] I reviewed the [Discussions](https://github.com/ggml-org/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

**tl;dr:** llama-perplexity currently allocates system memory in the order of ``O(ctx * vocab)``. However, it is possible to accomplish the same with only ``O(vocab)``, which makes large context evaluations significantly more practical.

### Motivation

I am exploring [KL divergence for Qwen3.8 27B](https://designprincipia.com/post/qwen3.8-weight-quantization-ladder/) on consumer hardware (I am limited to 32GB of system memory).

Beyond ~20-30K context size, it is effectively impractical for me to compute the logit distribution baseline file or the KLD. I need to set up a large swap file, and it takes around 1-2 hours to evaluate a run at 90K context size.

The proposed implementation below allows one to accomplish the same with orders of magnitude less system memory. In my particular case, it also drops runtime from hours to ~5 minutes by avoiding disk swap.

### Currently how it works:
 
1. Currently llama-perplexity allocates a vector to store logit values per vocabulary entry for all the tokens in the context window (for a given chunk).  i.e ``O(ctx * vocab)``
2. Then it batch decodes the entire corpus and stores the logits into the vector
3. Then it process the entire vector and generates the baseline file (in case of --kl-divergence-base), or reads the entire baseline file and compares against the entire decoded logits to calculate the KLD metrics.

### Idea/Proposal:

1. Instead of allocating a buffer for all tokens * vocab, only allocate a buffer for the size of tokens in the batch, so effectively O(batch * vocab) ~= O(vocab)
2. As each batch decode output is read from VRAM into system ram, process that right away instead of keeping it around in memory. That means either writing to disk (--kl-divergence-base) or reading from disk and calculating the KLD
3.  After each batch is process, then we can reuse the system ram buffer, rather than allocating more.

### Implementation:

I implemented the idea below on this [branch](https://github.com/ggml-org/llama.cpp/compare/master...andreesteve:llama.cpp:llama-perplexity-memory-optimization?expand=1#diff-22563b87e2d33ad7ab87078db474a7faecc9e8be63074c85c05e1e84e8740f63) with AI assistance. I am following the contribution guidelines to gather support before sending a PR.
